api/hypershift/v1beta1/hostedcluster_types.go

```diff
@@ -351,141 +351,161 @@ const (
+	// +kubebuilder:validation:XValidation:rule="self.all(s, !(s.service == 'APIServer' && s.servicePublishingStrategy.type == 'Route') || has(s.servicePublishingStrategy.route.hostname))",message="If serviceType is 'APIServer' and publishing strategy is 'Route', then hostname must be set"
+	// +kubebuilder:validation:XValidation:rule="['APIServer', 'OAuthServer', 'Konnectivity', 'Ignition'].all(requiredType, self.exists(s, s.service == requiredType))",message="Services list must contain at least 'APIServer', 'OAuthServer', 'Konnectivity', and 'Ignition' service types"
+	// +kubebuilder:validation:XValidation:rule="self.filter(s, s.servicePublishingStrategy.type == 'Route' && has(s.servicePublishingStrategy.route) && has(s.servicePublishingStrategy.route.hostname)).all(x, self.filter(y, y.servicePublishingStrategy.type == 'Route' && (has(y.servicePublishingStrategy.route) && has(y.servicePublishingStrategy.route.hostname) && y.servicePublishingStrategy.route.hostname == x.servicePublishingStrategy.route.hostname)).size() <= 1)",message="Each nodePort publishingStrategy 'nodePort' and 'hostname', and each route publishingStrategy 'hostname' must be unique within the Services list."
+	// +kubebuilder:validation:XValidation:rule="self.filter(s, s.servicePublishingStrategy.type == 'NodePort' && has(s.servicePublishingStrategy.nodePort) && has(s.servicePublishingStrategy.nodePort.address) && has(s.servicePublishingStrategy.nodePort.port)).all(x, self.filter(y, y.servicePublishingStrategy.type == 'NodePort' && (has(y.servicePublishingStrategy.nodePort) && has(y.servicePublishingStrategy.nodePort.address) && y.servicePublishingStrategy.nodePort.address == x.servicePublishingStrategy.nodePort.address && has(y.servicePublishingStrategy.nodePort.port) && y.servicePublishingStrategy.nodePort.port == x.servicePublishingStrategy.nodePort.port )).size() <= 1)",message="Each nodePort publishingStrategy 'nodePort' and 'hostname', and each route publishingStrategy 'hostname' must be unique within the Services list."
+	// +immutable
 	Services []ServicePublishingStrategyMapping `json:"services"`
 
-	// PullSecret references a pull secret to be injected into the container
-	// runtime of all cluster nodes. The secret must have a key named
-	// ".dockerconfigjson" whose value is the pull secret JSON.
+	// pullSecret is a local reference to a Secret that must have a ".dockerconfigjson" key whose content must be a valid Openshift pull secret JSON.
+	// If the reference is set but none of the above requirements are met, the HostedCluster will enter a degraded state.
+	// TODO(alberto): Signal this in a condition.
+	// This pull secret will be part of every payload generated by the controllers for any NodePool of the HostedCluster
+	// and it will be injected into the container runtime of all NodePools.
+	// Changing this value will trigger a rollout for all existing NodePools in the cluster.
+	// +required
 	PullSecret corev1.LocalObjectReference `json:"pullSecret"`
```

Comment:
Prefer to use our own types, instead of importing, it means we can tailor the output, and, it means that if the external type adds a new field, we don't magically gain that in our API, as a new unimplemented field

Reply:
Thanks for pointing this out, yes this is tech debt. I'll revisit this in a different PR, this iteration is getting big enough